sycl/doc/design/SYCLBINDesign.md

```diff
@@ -0,0 +1,287 @@
+name "SYCLBIN/ir module metadata", as described in the
+[PropertySets.md](PropertySets.md#syclbinir-module-metadata) design document.
+
+
+#### Native device code image
+
+An native device code image contains the binary data for the corresponding
+module AOT compiled for a specific device, identified by the architecture
+string.  The runtime library will attempt to map these to the architecture
+enumerators in the
+[sycl_ext_oneapi_device_architecture](../extensions/experimental/sycl_ext_oneapi_device_architecture.asciidoc)
+extension.
+
+
+##### Native device code image header
+
+A native device code image header contains the following fields in the stated
+order:
+
+| Type       | Description                                                                         |
+| ---------- | ----------------------------------------------------------------------------------- |
+| `uint64_t` | Byte offset of the metadata from the start of the metadata byte table.              |
+| `uint64_t` | Byte size of the metadata in the metadata byte table.                               |
+| `uint64_t` | Byte offset of the device code image bytes from the start of the binary byte table. |
+| `uint64_t` | Byte size of the device code image bytes in the binary byte table.                  |
+
+
+##### Native device code image metadata
+
+A native device code image metadata entry contains a single property set with
+the identifying name "SYCLBIN/native device code image module metadata", as
+described in the
+[PropertySets.md](PropertySets.md#syclbinnative-device-code-image-metadata)
+design document.
+
+
+### Byte tables
+
+A byte table contains dynamic data, such as metadata and binary blobs. The
+contents of it is generally referenced by an offset specified in the headers.
+
+
+### SYCLBIN version changelog
+
+The SYCLBIN format is subject to change, but any such changes must come with an
+increment to the version number in the header and a subsection to this section
+describing the change.
+
+Additionally, any changes to the property set structure that affects the way the
+runtime has to parse the contained property sets will require an increase in the
+SYCLBIN version. Adding new property set names or new predefined properties only
+require a SYCLBIN version change if the the SYCLBIN consumer cannot safely
+ignore the property.
+
+
+#### Version 1
+
+ * Initial version of the layout.
+
+
+## Clang driver changes
+
+The clang driver needs to accept the following new flags:
+
+<table>
+<tr>
+<th>Option</th>
+<th>Description</th>
+</tr>
+<tr>
+<td>`-fsyclbin`</td>
+<td>
+If this option is set, the output of the invocation is a SYCLBIN file with the
+.syclbin file extension. This skips the host-compilation invocation of the typical
+`-fsycl` pipeline, instead passing the output of the clang-offload-packager
+invocation to clang-linker-wrapper together with the new `--syclbin` flag.
+
+Setting this option implies `-fsycl` and `-fsycl-device-only`.
```

Comment:
Now that I see this written, what is the expectation of behavior when `-fsyclbin -fsycl-device-only` is used on the command line?  The natural behavior here is for `-fsycl-device-only` to be seen and taken advantage of, which will create the device only file and stop.  `-fsyclbin` does additional work after the device compilation.

Reply:
That's a good point. Maybe it would be better to issue a diagnostic if they are used together? Seems like they may not be totally related, despite there being some conceptual overlap.

Reply:
I think the standard 'unused argument' diagnostic should be sufficient when used together:
```
> clang++ --offload-new-driver -fsycl-device-only -fsyclbin ~/a.cpp
clang++: warning: argument unused during compilation: '-fsyclbin' [-Wunused-command-line-argument]
```

Reply:
Sounds good! I've reworded this part to explicitly say that the `-fsyclbin` will be unused.